Connectivity handler

ancaantochi opened the pull request:
handles connectivity event from upstream and sends events to local broker 

(Contains changes in mqtt3 client)

ancaantochi committed:
Version 1

ancaantochi committed:
Version 2

ancaantochi committed:
Client send disconnect event

ancaantochi committed:
Client

ancaantochi committed:
New version

ancaantochi committed:
Merge branch 'andsmi/bridge-upstream-pump' of https://github.com/and-rewsmith/iotedge into ancan/bridge-connectivity-pump

ancaantochi committed:
Fixes

ancaantochi committed:
authz all

ancaantochi committed:
authz all

ancaantochi committed:
add log to trace

ancaantochi committed:
add log to trace

ancaantochi committed:
rename topic

ancaantochi committed:
Merge branch 'andsmi/bridge-upstream-pump' of https://github.com/and-rewsmith/iotedge into ancan/bridge-connectivity-pump

ancaantochi committed:
fix merge

ancaantochi committed:
clean up

ancaantochi committed:
Merge branch 'andsmi/bridge-upstream-pump' of https://github.com/and-rewsmith/iotedge into ancan/bridge-connectivity-pump

ancaantochi committed:
Tests

ancaantochi committed:
clean up

ancaantochi committed:
Merge branch 'andsmi/bridge-upstream-pump' of https://github.com/and-rewsmith/iotedge into ancan/bridge-connectivity-pump

ancaantochi committed:
Fix merge

ancaantochi committed:
Clean up

ancaantochi committed:
Revert "authz all"

This reverts commit 0a8009ce05db5bc54afbea887f0e371b39e52e83.

ancaantochi committed:
Revert "authz all"

This reverts commit d88acbf402e67a87e7b0e6328f3b2693fc0c8044.

dmolokanov commented on mqtt/mqtt-bridge/src/connectivity_handler.rs:
```diff
@@ -0,0 +1,177 @@
+
+use crate::bridge::{BridgeError, BridgeMessage, ConnectivityState};
+
+impl ToString for ConnectivityState {
```
usually in rust you don't need to impl ToString, Display instead. ToString is auto-implemented when thet ype impl Display

ancaantochi replied:
addressed in master PR

dmolokanov commented on mqtt/mqtt-bridge/src/connectivity_handler.rs:
```diff
@@ -0,0 +1,177 @@
+    sender: UnboundedSender<BridgeMessage>,
+}
+
+impl ConnectivityHandler {
```
lets make ConnectivityHandler to impl EventHandler?

ancaantochi replied:
addressed in master PR

dmolokanov commented on mqtt/mqtt-bridge/src/lib.rs:
```diff
@@ -10,9 +10,11 @@
     clippy::must_use_candidate,
     clippy::missing_errors_doc
 )]
+#![recursion_limit = "256"]
```
this looks weird

ancaantochi replied:
this is because of loop with select! in pump. I added the loop because the future might be completed when there is no connectivity receiver in the pump

dmolokanov replied:
can we switch it to the `future::select` maybe?

dmolokanov commented on mqtt/mqtt-bridge/src/pump.rs:
```diff
@@ -6,19 +6,21 @@ use futures_util::{
     persist::{MessageLoader, PublicationStore, WakingMemoryStore},
 };
 
+const CONNECTIVITY_TOPIC: &str = "$internal/connectivity";
```
We need to check with @vipeller to see what topic we should use for this event. `$edgehub/bridge/connected` or something like this. Also we may want to make it retained message

ancaantochi replied:
I checked with him and he said I could pick a topic, if $edgehub makes more sense I can change.

dmolokanov commented on mqtt/mqttd/src/tracing/edgehub.rs:
```diff
@@ -13,11 +14,16 @@ pub fn init() {
         .or_else(|_| EnvFilter::try_from_default_env())
         .unwrap_or_else(|_| EnvFilter::new("info"));
 
+    let filter = log::LevelFilter::from_str(log_level.to_string().as_str())
```
You say it captures log entries, but then it publishes it as trace!() in tracing?

ancaantochi replied:
it captures logs as tracing and if the level is all lowercase it works as expected. The issue is when level is Debug for example it will log everything as trace, but the issue is not because of this change, it seems to be in EnvFilter

dmolokanov commented on mqtt/mqtt-bridge/src/bridge.rs:
```diff
@@ -74,11 +88,17 @@ impl Bridge {
+
+        // component_sender can be used for configuration changes
+        let (connectivity_sender, connectivity_receiver) =
+            mpsc::unbounded_channel::<BridgeMessage>();
```
not sure that we want unbounded channel here. what do you think?

ancaantochi replied:
addressed in master PR

dmolokanov commented on mqtt/mqtt-bridge/src/bridge.rs:
```diff
@@ -105,6 +126,7 @@ impl Bridge {
             local_subscriptions,
             incoming_loader,
             incoming_persist,
+            Some(connectivity_receiver),
```
it is not for connectivity events anymore. should we make it more generic?

dmolokanov commented on mqtt/mqtt-bridge/src/bridge.rs:
```diff
@@ -74,11 +88,17 @@ impl Bridge {
             .collect::<Result<Vec<_>, _>>()?;
+
+        // component_sender can be used for configuration changes
+        let (connectivity_sender, connectivity_receiver) =
```
I wonder if we can move channel creation inside pump. wrap sender part into PumpHandle and pass PumpHandle everywhere instead? Pattern is similar what we have in the Broker struct

dmolokanov commented on mqtt/mqtt-bridge/src/bridge.rs:
```diff
@@ -16,6 +18,18 @@ use crate::{
 const BATCH_SIZE: usize = 10;
 
+#[derive(Debug, PartialEq)]
+pub enum BridgeMessage {
```
this is the message to pump, right? should we call it PumpMessage instead?

ancaantochi replied:
addressed in master PR

dmolokanov committed:
[mqtt] RPC command handler (part 1) (#3687)

The first part of the RPC command handler work.
This part contains command definition, deserialization from BSON, and the command handler itself.
Exact integration with Bridge infra will be is separate PR.

huguesBouvier committed:
Fixing clippy iotedged (#3696)

* Fix clipppy

* Fixing clippy

* Fixing review comments

* fixing forgotten map_or_else

* fix clippy test

Co-authored-by: Ubuntu <iotedgeuser@hug3l3edge.0x1kslxyg1putjewhwmoznt1le.yx.internal.cloudapp.net>
Co-authored-by: Ubuntu <azureuser@HuG1.2bw1qssgs53udfdpqpfa4uhcsd.yx.internal.cloudapp.net>

huguesBouvier committed:
Workload api can take hostname with capital letter (#3683)

* Fixing bug in edged

* Fixing iotedge check for IP

* disabling parent hostname container test for IP

* clippy

* fix clippy due to clippy version upgrade

* rolling back clippy changes

* Fix review comments

* fixing cargo fmt

Co-authored-by: Ubuntu <iotedgeuser@hug3l3edge.0x1kslxyg1putjewhwmoznt1le.yx.internal.cloudapp.net>

richma-ms committed:
Remove deployment manifest minor version strict validation (#3692) (#3698)

Remove deployment manfiest minor version strict validation

and-rewsmith committed:
address self comments

philipktlin committed:
Create a new api-proxy-module pipeline for release (#3677)

Create a new pipeline to release api-proxy-module docker image.

and-rewsmith committed:
remove TODO REVIEW

and-rewsmith committed:
[mqtt] Add todo for persisted store (#3691)

We need to handle offset correctly for persisted storage. I don't want to forget so am adding this note.

varunpuranik committed:
Fix quantiles for Duration and Timer to align with the quantiles for histogram (#3670)

We want to use  0.1, 0.5, 0.9, 0.99 quantiles for all our metrics. This was updated for Histogram metric in this commit https://github.com/Azure/iotedge/commit/f85b8389f2c6366a97476c13b336afe7050a79f8 but was not updated for Duration/Timer. This PR takes care of that.

ancaantochi committed:
Merge branch 'andsmi/bridge-upstream-pump' of https://github.com/and-rewsmith/iotedge into ancan/bridge-connectivity-pump

and-rewsmith committed:
pump context struct

and-rewsmith committed:
log pump context in client

and-rewsmith committed:
refactor bridge new

pmzara committed:
Fix iotedge support-bundle doesn't work if iotedged isn't running (#3704)

Currently iotedge support-bundle will fail if iotedged isn't running. It should still collect host and docker logs.
For this fix, I'm catching the failed operation and return a empty list. This will not cause any problem in the execution of the functions and will produce the logs for all the others operations except for the operations that using the information produced by the runtime.

ancaantochi committed:
mqtt client send disconnect event (#3653)

mqtt3 client to send an event when it disconnects from server.

and-rewsmith committed:
merge conflict upstream master

dmolokanov committed:
[mqtt] Revert mockall version to the one from crates.io (#3712)

Revert mockall version to the one from crates.io

and-rewsmith committed:
merge conflict upstream master

ancaantochi committed:
Add connectivity handler (#3699)

Adding handler for connectivity events. 
The idea is to create the channel in the bridge and the sender is passed to upstream pump and receiver to downstream pump. This will be in a different PR since changes are based on the pump.

Note: PR has to be merged after changes to mqtt3 client changes.

and-rewsmith committed:
add doc comments to pump abstractions

and-rewsmith committed:
pumps can only be created in pairs

and-rewsmith committed:
add comment explaining bridge controller behavior

and-rewsmith committed:
merge conflict upstream master

and-rewsmith committed:
fix clippy

ancaantochi committed:
Merge branch 'andsmi/bridge-upstream-pump-PR' of https://github.com/and-rewsmith/iotedge into ancan/bridge-connectivity-pump
